File: packages/protopy/agent/bifrost.py
# bifrost is the protofy protocol for 
# agents connection and comunication.

# brifrost actions
#  - register: a new agent (agents/<agent_name>/register)

# bifrost defaults
#  - actions endpoint: agents/<agent_name>/subsystem/<subsystem_name>/action/<action_name>
#  - monitors endpoint: agents/<agent_name>/subsystem/<subsystem_name>/monitor/<monitor_name>

import json

# actions
def register(publisher, name, subsytems): 
    return publisher("agents/"+name+"/register", json.dumps({ "subsystems": subsytems }))

def pub_monitor(publisher, name, subsystem_name, monitor_name, value): 
    return publisher(gen_monitor_endpoint(name, subsystem_name, monitor_name), json.dumps({ "value": value }))

# endpoints
def gen_action_endpoint(name, subsystem_name, action_name): 
    return "agents/"+name+"/subsystem/"+subsystem_name+"/action/"+action_name

def gen_monitor_endpoint(name, subsystem_name, monitor_name): 
    return "agents/"+name+"/subsystem/"+subsystem_name+"/monitor/"+monitor_name

File: packages/agents/js/bifrost.ts
// Bifrost is the protocol for agents connection and communication.

// Bifrost actions:
//  - register: a new agent (agents/<agent_name>/register)

// Bifrost defaults:
//  - actions endpoint: agents/<agent_name>/subsystem/<subsystem_name>/action/<action_name>
//  - monitors endpoint: agents/<agent_name>/subsystem/<subsystem_name>/monitor/<monitor_name>

// actions
export function register(publisher, name, subsystems) {
  return publisher(`agents/${name}/register`, JSON.stringify({ subsystems: subsystems }));
}

export function pubMonitor(publisher, name, subsystemName, monitorName, value) {
  return publisher(genMonitorEndpoint(name, subsystemName, monitorName), JSON.stringify({ value: value }));
}

// endpoints
export function genActionEndpoint(name, subsystemName, actionName) {
  return `agents/${name}/subsystem/${subsystemName}/action/${actionName}`;
}

export function genMonitorEndpoint(name, subsystemName, monitorName) {
  return `agents/${name}/subsystem/${subsystemName}/monitor/${monitorName}`;
}

